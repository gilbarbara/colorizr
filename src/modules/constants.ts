import { ColorKeysTuple, ColorModelKey } from '~/types';

export const COLOR_KEYS: Record<ColorModelKey, ColorKeysTuple> = {
  hsl: ['h', 's', 'l'],
  oklab: ['l', 'a', 'b'],
  oklch: ['l', 'c', 'h'],
  rgb: ['r', 'g', 'b'],
};
export const COLOR_MODELS: ColorModelKey[] = ['hsl', 'oklab', 'oklch', 'rgb'];
export const DEG2RAD = Math.PI / 180;
export const LAB_TO_LMS = {
  l: [0.3963377773761749, 0.2158037573099136],
  m: [-0.1055613458156586, -0.0638541728258133],
  s: [-0.0894841775298119, -1.2914855480194092],
};
export const LRGB_TO_LMS = {
  l: [0.4122214708, 0.5363325363, 0.0514459929],
  m: [0.2119034982, 0.6806995451, 0.1073969566],
  s: [0.0883024619, 0.2817188376, 0.6299787005],
};
export const LSM_TO_LAB = {
  l: [0.2104542553, 0.793617785, 0.0040720468],
  a: [1.9779984951, 2.428592205, 0.4505937099],
  b: [0.0259040371, 0.7827717662, 0.808675766],
};
export const LSM_TO_RGB = {
  r: [4.0767416360759583, -3.3077115392580629, 0.2309699031821043],
  g: [-1.2684379732850315, 2.6097573492876882, -0.341319376002657],
  b: [-0.0041960761386756, -0.7034186179359362, 1.7076146940746117],
};
export const PRECISION = 5;
export const RAD2DEG = 180 / Math.PI;

export const MESSAGES = {
  alpha: 'amount must be a number between 0 and 1',
  left: 'left is required and must be a string',
  right: 'right is required and must be a string',
  input: 'input is required',
  inputHex: 'input is required and must be a hex',
  inputNumber: 'input is required and must be a number',
  inputString: 'input is required and must be a string',
  invalid: 'invalid input',
  invalidCSS: 'invalid CSS string',
  options: 'invalid options',
  threshold: 'threshold must be a number between 0 and 255',
} as const;
