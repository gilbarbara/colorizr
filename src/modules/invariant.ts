export function invariant(condition: boolean, message: string): asserts condition {
  if (condition) {
    return;
  }

  /* istanbul ignore else */
  if (process.env.NODE_ENV !== 'production') {
    if (message === undefined) {
      throw new Error('invariant requires an error message argument');
    }
  }

  const error = !message
    ? new Error(
        'Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.',
      )
    : new Error(message);

  error.name = 'colorizr';

  throw error;
}
