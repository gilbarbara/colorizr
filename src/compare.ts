import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { isString } from '~/modules/validators';

import getBrightnessDifference from '~/brightness-difference';
import getColorDifference from '~/color-difference';
import getContrast from '~/contrast';
import { Analysis } from '~/types';

/**
 * Check 2 colors for WCAG compliance.
 */
export default function compare(left: string, right: string): Analysis {
  invariant(isString(left), MESSAGES.left);
  invariant(isString(right), MESSAGES.right);

  const colorThreshold = 500;
  const brightnessThreshold = 125;

  const colorDifference = getColorDifference(left, right);
  const contrast = getContrast(left, right);
  const brightnessDifference = getBrightnessDifference(left, right);

  const isBright = brightnessDifference >= brightnessThreshold;
  const hasEnoughDifference = colorDifference >= colorThreshold;

  let compliant = 0;

  if (isBright && hasEnoughDifference) {
    compliant = 2;
  } else if (isBright || hasEnoughDifference) {
    compliant = 1;
  }

  return {
    brightnessDifference,
    colorDifference,
    compliant,
    contrast,
    largeAA: contrast >= 3,
    largeAAA: contrast >= 4.5,
    normalAA: contrast >= 4.5,
    normalAAA: contrast >= 7,
  };
}
