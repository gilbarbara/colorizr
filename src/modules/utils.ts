import { COLOR_KEYS, COLOR_MODELS, MESSAGES, PRECISION } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import {
  isHSL,
  isLAB,
  isLCH,
  isNumber,
  isPlainObject,
  isRGB,
  isValidColorModel,
} from '~/modules/validators';

import {
  Alpha,
  ColorModel,
  ColorModelKey,
  ColorModelKeys,
  ConverterParameters,
  LAB,
  LCH,
  PlainObject,
} from '~/types';

export function addAlpha<T extends ColorModel>(input: any, alpha?: Alpha): T {
  invariant(isValidColorModel(input), MESSAGES.invalid);

  let value = alpha;

  if (!value) {
    return input;
  }

  /* c8 ignore next 3 */
  if (value > 1) {
    value /= 100;
  }

  if (value === 1) {
    return input;
  }

  return { ...input, alpha: value };
}

/**
 * Clamp a value between a min and max
 * @param value
 * @param [min=0] - The minimum value
 * @param [max=100] - The maximum value
 */
export function clamp(value: number, min = 0, max = 100) {
  return Math.min(Math.max(value, min), max);
}

/**
 * Constrain the degrees between 0 and 360
 */
export function constrainDegrees(input: number, amount: number): number {
  invariant(isNumber(input), MESSAGES.inputNumber);

  let value = input + amount;

  if (value > 360) {
    value %= 360;
  }

  if (value < 0) {
    value += 360;
  }

  return Math.abs(value);
}

/**
 * Limit values per type.
 */
export function limit<TModel extends Extract<ColorModelKey, 'hsl' | 'rgb'>>(
  input: number,
  model: TModel,
  key: ColorModelKeys<TModel>,
): number {
  invariant(isNumber(input), 'Input is not a number');
  invariant(COLOR_MODELS.includes(model), `Invalid model${model ? `: ${model}` : ''}`);
  invariant(COLOR_KEYS[model].includes(key), `Invalid key${key ? `: ${key}` : ''}`);

  switch (model) {
    case 'hsl': {
      invariant(COLOR_KEYS.hsl.includes(key), 'Invalid key');

      if (['s', 'l'].includes(key)) {
        return clamp(input);
      }

      return clamp(input, 0, 360);
    }
    case 'rgb': {
      invariant(COLOR_KEYS.rgb.includes(key), 'Invalid key');

      return clamp(input, 0, 255);
    }
    /* c8 ignore next 3 */
    default: {
      throw new Error('Invalid inputs');
    }
  }
}

/**
 * Parse the input parameters
 */
export function parseInput<T extends ColorModel>(
  input: ConverterParameters<T>,
  model: ColorModelKey,
): T {
  const keys = COLOR_KEYS[model];
  const validator = {
    hsl: isHSL,
    oklab: isLAB,
    oklch: isLCH,
    rgb: isRGB,
  };

  invariant(isPlainObject(input) || Array.isArray(input), MESSAGES.invalid);

  const value = Array.isArray(input)
    ? ({ [keys[0]]: input[0], [keys[1]]: input[1], [keys[2]]: input[2] } as unknown as T)
    : input;

  invariant(validator[model](value), `invalid ${model} color`);

  return value;
}

/**
 * Creates an object composed of the picked source properties.
 */
export function pick(input: PlainObject, options: string[]): PlainObject {
  if (!Array.isArray(options)) {
    throw new TypeError('options must be an array');
  }

  return options
    .filter(d => typeof input[d] !== 'undefined')
    .reduce((acc: PlainObject, d) => {
      acc[d] = input[d];

      return acc;
    }, {});
}

/**
 * Restrict the values to a certain number of digits.
 */
export function restrictValues<T extends LAB | LCH>(
  input: T,
  precision: number = PRECISION,
  forcePrecision = true,
): T {
  const output = new Map(Object.entries(input));

  for (const [key, value] of output.entries()) {
    output.set(key, round(value, precision, forcePrecision));
  }

  return Object.fromEntries(output) as T;
}

/**
 * Round decimal numbers.
 */
export function round(input: number, precision = 2, forcePrecision = true): number {
  if (!isNumber(input) || input === 0) {
    return 0;
  }

  if (forcePrecision) {
    const factor = 10 ** precision;

    return Math.round(input * factor) / factor;
  }

  const absInput = Math.abs(input);

  let digits = Math.abs(Math.ceil(Math.log(absInput) / Math.LN10));

  if (digits === 0) {
    digits = 2;
  } else if (digits > precision) {
    digits = precision;
  }

  let exponent = precision - (digits < 0 ? 0 : digits);

  if (exponent <= 1 && precision > 1) {
    exponent = 2;
  } else if (exponent > precision || exponent === 0) {
    exponent = precision;
  }

  const factor = 10 ** exponent;

  return Math.round(input * factor) / factor;
}
