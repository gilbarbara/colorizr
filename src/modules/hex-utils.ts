import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { round } from '~/modules/utils';
import { isHex, isNumber } from '~/modules/validators';

import { Alpha } from '~/types';

/**
 * Add an alpha value to a hex string
 */
export function addAlphaToHex(input: string, alpha: Alpha): string {
  invariant(isHex(input), MESSAGES.inputHex);
  invariant(isNumber(alpha), MESSAGES.inputNumber);

  if (alpha >= 1) {
    return removeAlphaFromHex(input);
  }

  return `${removeAlphaFromHex(input)}${convertAlphaToHex(alpha)}`;
}

/**
 * Convert an alpha value to a hex value.
 */
export function convertAlphaToHex(input: Alpha): string {
  invariant(isNumber(input), MESSAGES.inputNumber);

  let alpha = input;

  if (input > 1) {
    alpha /= 100;
  }

  return Math.round(alpha * 255)
    .toString(16)
    .padStart(2, '0');
}

/**
 * Extract the alpha value from a hex string
 */
export function extractAlphaFromHex(input: string): number {
  invariant(isHex(input), MESSAGES.inputString);

  const alpha = input.substring(7, 9);

  if (!alpha) {
    return 1;
  }

  return round(parseInt(alpha, 16) / 255);
}

export function hexadecimalToNumber(input: string) {
  return round(parseInt(input, 16));
}

/**
 * Remove the alpha value from a hex string
 */
export function removeAlphaFromHex(input: string) {
  invariant(isHex(input), MESSAGES.inputHex);

  if (input.length === 5) {
    return input.substring(0, 4);
  }

  return input.substring(0, 7);
}
