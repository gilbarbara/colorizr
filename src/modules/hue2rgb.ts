import { invariant, isNumber, round } from './utils';

/**
 * Convert hue to RGB using chroma and median point
 */
export default function hue2rgb(point: number, chroma: number, h: number): number {
  invariant(isNumber(point) && isNumber(chroma) && isNumber(h), 'point, chroma and h are required');
  let hue = h;

  if (hue < 0) {
    hue += 1;
  }

  if (hue > 1) {
    hue -= 1;
  }

  if (hue < 1 / 6) {
    return round(point + (chroma - point) * 6 * hue, 4);
  }

  if (hue < 1 / 2) {
    return round(chroma, 4);
  }

  if (hue < 2 / 3) {
    return round(point + (chroma - point) * (2 / 3 - hue) * 6, 4);
  }

  return round(point, 4);
}
