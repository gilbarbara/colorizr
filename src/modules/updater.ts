import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { clamp } from '~/modules/utils';
import { isHex, isNamedColor, isNumber, isString } from '~/modules/validators';

import extractColorParts from '~/extract-color-parts';
import formatCSS from '~/format-css';
import parseCSS from '~/parse-css';
import { ColorModelKeys, ColorType } from '~/types';

/**
 * Update color properties
 */
export default function updater(
  key: ColorModelKeys<'hsl'>,
  operator: '+' | '-',
  format?: ColorType,
) {
  return (input: string, amount: number) => {
    invariant(isString(input), MESSAGES.inputString);
    invariant(isNumber(amount), MESSAGES.alpha);

    const color = parseCSS(input, 'hsl');
    const output = isHex(input) || isNamedColor(input) ? 'hex' : extractColorParts(input).model;

    return formatCSS(
      {
        ...color,
        [key]: clamp(color[key] + (operator === '+' ? amount : -amount), 0, 100),
      },
      { format: format ?? output },
    );
  };
}
