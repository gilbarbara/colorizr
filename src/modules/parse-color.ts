import { MESSAGES } from '~/modules/constants';
import { addAlphaToHex } from '~/modules/hex-utils';
import { invariant } from '~/modules/invariant';
import { addAlpha, limit } from '~/modules/utils';
import { isHex, isHSL, isLAB, isLCH, isPlainObject, isRGB, isString } from '~/modules/validators';

import * as converters from '~/converters';
import extractColorParts from '~/extract-color-parts';
import parseCSS from '~/parse-css';
import { Colors, HSL, LAB, LCH, PlainObject, RGB } from '~/types';

export default function parseColor(color: string | HSL | LAB | LCH | RGB): Colors {
  invariant(!!color, MESSAGES.input);

  const output: PlainObject = {};

  if (isString(color)) {
    const { alpha = 1 } = extractColorParts(color);
    const type = isHex(color) ? 'hex' : extractColorParts(color).model;

    output.hex = addAlphaToHex(parseCSS(color, 'hex'), alpha);
    output.hsl = addAlpha(parseCSS(color, 'hsl'), alpha);
    output.oklab = addAlpha(parseCSS(color, 'oklab'), alpha);
    output.oklch = addAlpha(parseCSS(color, 'oklch'), alpha);
    output.rgb = addAlpha(parseCSS(color, 'rgb'), alpha);

    output.alpha = alpha;
    output.type = type;
  } else if (isPlainObject(color)) {
    const { alpha = 1 } = color;

    if (isHSL(color)) {
      output.hsl = {
        h: limit(color.h, 'hsl', 'h'),
        s: limit(color.s, 'hsl', 's'),
        l: limit(color.l, 'hsl', 'l'),
      };
      output.rgb = converters.hsl2rgb(output.hsl);
      output.oklab = converters.hsl2oklab(output.hsl);
      output.oklch = converters.hsl2oklch(output.hsl);
      output.type = 'hsl';
    } else if (isLAB(color)) {
      output.hsl = converters.oklab2hsl(color);
      output.oklab = color;
      output.oklch = converters.oklab2oklch(color);
      output.rgb = converters.oklab2rgb(color);
      output.type = 'oklab';
    } else if (isLCH(color)) {
      output.hsl = converters.oklch2hsl(color);
      output.oklab = converters.oklch2oklab(color);
      output.oklch = color;
      output.rgb = converters.oklch2rgb(color);
      output.type = 'oklch';
    } else if (isRGB(color)) {
      output.rgb = {
        r: limit(color.r, 'rgb', 'r'),
        g: limit(color.g, 'rgb', 'g'),
        b: limit(color.b, 'rgb', 'b'),
      };
      output.hsl = converters.rgb2hsl(output.rgb);
      output.oklab = converters.rgb2oklab(output.rgb);
      output.oklch = converters.rgb2oklch(output.rgb);
      output.type = 'rgb';
    } else {
      throw new Error('invalid color');
    }

    output.hex = addAlphaToHex(converters.hsl2hex(output.hsl), alpha);
    output.hsl = addAlpha(output.hsl, alpha);
    output.oklab = addAlpha(output.oklab, alpha);
    output.oklch = addAlpha(output.oklch, alpha);
    output.rgb = addAlpha(output.rgb, alpha);

    output.alpha = alpha;
  } else {
    throw new Error(MESSAGES.input);
  }

  return output as Colors;
}
