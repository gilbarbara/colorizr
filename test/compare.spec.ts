import { MESSAGES } from '~/modules/constants';

import compare from '~/compare';

describe('compare', () => {
  it.each([
    ['#ff0044', 'rgb(221, 0, 255)'],
    ['rgb(255, 0, 68)', '#00ffbb'],
    ['hsl(344, 100%, 50%)', 'oklch(1 0 0)'],
    ['#ff0044', 'oklab(0 0 0)'],
    ['#fff', 'rgb(119, 119, 119)'],
    ['oklab(0 0 0)', 'oklch(1 0 0)'],
  ])('%s with %s should return an analysis', (left, right) => {
    expect(compare(left, right)).toMatchSnapshot();
  });

  it('should fail with invalid parameters', () => {
    // @ts-expect-error - no parameters
    expect(() => compare()).toThrow(MESSAGES.left);
    // @ts-expect-error - invalid parameters
    expect(() => compare([])).toThrow(MESSAGES.left);
    // @ts-expect-error - invalid parameters
    expect(() => compare('#f04', [])).toThrow(MESSAGES.right);
  });
});
