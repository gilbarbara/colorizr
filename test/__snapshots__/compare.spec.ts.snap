// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compare > #ff0044 with oklab(0 0 0) should return an analysis 1`] = `
{
  "brightnessDifference": 83.997,
  "colorDifference": 323,
  "compliant": 0,
  "contrast": 5.34,
  "largeAA": true,
  "largeAAA": true,
  "normalAA": true,
  "normalAAA": false,
}
`;

exports[`compare > #ff0044 with rgb(221, 0, 255) should return an analysis 1`] = `
{
  "brightnessDifference": 11.152,
  "colorDifference": 221,
  "compliant": 0,
  "contrast": 1.03,
  "largeAA": false,
  "largeAAA": false,
  "normalAA": false,
  "normalAAA": false,
}
`;

exports[`compare > #fff with rgb(119, 119, 119) should return an analysis 1`] = `
{
  "brightnessDifference": 136,
  "colorDifference": 408,
  "compliant": 1,
  "contrast": 4.48,
  "largeAA": true,
  "largeAAA": false,
  "normalAA": false,
  "normalAAA": false,
}
`;

exports[`compare > hsl(344, 100%, 50%) with oklch(1 0 0) should return an analysis 1`] = `
{
  "brightnessDifference": 171.003,
  "colorDifference": 442,
  "compliant": 1,
  "contrast": 3.94,
  "largeAA": true,
  "largeAAA": false,
  "normalAA": false,
  "normalAAA": false,
}
`;

exports[`compare > oklab(0 0 0) with oklch(1 0 0) should return an analysis 1`] = `
{
  "brightnessDifference": 255,
  "colorDifference": 765,
  "compliant": 2,
  "contrast": 21,
  "largeAA": true,
  "largeAAA": true,
  "normalAA": true,
  "normalAAA": true,
}
`;

exports[`compare > rgb(255, 0, 68) with #00ffbb should return an analysis 1`] = `
{
  "brightnessDifference": 87.006,
  "colorDifference": 629,
  "compliant": 1,
  "contrast": 3,
  "largeAA": true,
  "largeAAA": false,
  "normalAA": false,
  "normalAAA": false,
}
`;
